File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/value/AnimatedValue.java
package com.mcreater.fxfluent.controls.value;

import javafx.animation.KeyFrame;
import javafx.animation.KeyValue;
import javafx.animation.Timeline;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.util.Duration;

public class AnimatedValue<T> {
    public final ObjectProperty<T> property = new SimpleObjectProperty<>();
    private Timeline timeline;
    private final Duration duration;
    public AnimatedValue(T start, Duration duration) {
        property.set(start);
        this.duration = duration;
    }

    public void setValue(T value) {
        if (timeline != null) timeline.stop();
        property.set(value);
    }

    public void updateValue(T value) {
        if (timeline != null) timeline.stop();
        timeline = new Timeline(
                new KeyFrame(
                        Duration.ZERO,
                        new KeyValue(
                                property,
                                property.get()
                        )
                ),
                new KeyFrame(
                        duration,
                        new KeyValue(
                                property,
                                value
                        )
                )
        );
        timeline.setCycleCount(1);
        timeline.setAutoReverse(false);
        timeline.setDelay(Duration.ZERO);
        timeline.playFromStart();
    }
}

File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/value/AnimatedValue.kt
package com.mcreater.fxfluent.controls.value

import javafx.animation.Interpolator
import javafx.animation.KeyFrame
import javafx.animation.KeyValue
import javafx.animation.Timeline
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.util.Duration


class AnimatedValue<T>(start: T, duration: Duration, private val interpolator: Interpolator) {
    val property: ObjectProperty<T> = SimpleObjectProperty()
    private var timeline: Timeline? = null
    private val duration: Duration
    init {
        property.set(start)
        this.duration = duration
    }
    constructor(start: T, duration: Duration): this(start, duration, Interpolator.LINEAR)

    fun setValue(value: T) {
        if (timeline != null) timeline!!.stop()
        property.set(value)
    }

    fun updateValue(value: T) {
        if (timeline != null) timeline!!.stop()
        timeline = Timeline(
            KeyFrame(
                Duration.ZERO,
                KeyValue(
                    property,
                    property.get(),
                    interpolator
                )
            ),
            KeyFrame(
                duration,
                KeyValue(
                    property,
                    value,
                    interpolator
                )
            )
        )
        timeline!!.cycleCount = 1
        timeline!!.isAutoReverse = false
        timeline!!.delay = Duration.ZERO
        timeline!!.playFromStart()
    }
}
